### PR TITLE
models/entities: Add activities!

### PR DESCRIPTION
**Pull request by erincandescent**
Make users create a Create activity when created

### CODE REVIEW

File: models/entities/activity.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+	return api.Response{
+		Data: d,
+	}
+}
+
+func (self *Activity) Hydrate(ctx context.Context) (map[string]interface{}, error) {
+	if o, err := jsonld.Marshal(self); err == nil {
+		return jsonld.Compact(lib.GetHttpClient(ctx), o.(map[string]interface{}), "", "https://www.w3.org/ns/activitystreams")
+	} else {
```
Comment (liclac) [inline models/entities/activity.go]: No need for an `else` here

File: models/entities/activity.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+			Base: Base{
+				Meta: jsonld.Meta{},
+				ID:   id,
+				Type: types,
+			},
+		},
+	}
+	obj.SetSnowflake(flake)
+
+	if err := store.Insert(obj); err != nil {
```
Comment (liclac) [inline models/entities/activity.go]: You could `return obj, store.Insert(obj)`!

File: models/entities/activity.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+func (self *Activity) Hydrate(ctx context.Context) (map[string]interface{}, error) {
+	if o, err := jsonld.Marshal(self); err == nil {
+		return jsonld.Compact(lib.GetHttpClient(ctx), o.(map[string]interface{}), "", "https://www.w3.org/ns/activitystreams")
+	} else {
+		return nil, err
+	}
+}
+
+func NewActivity(store *Store, parent Entity, types []string) (*Activity, error) {
```
Comment (liclac) [inline models/entities/activity.go]: What would `parent` be in this case? I'm guessing for a Create event for a Note, the parent would be the Person who made it?

File: models/entities/person.go [hunk 4]
```diff
@@ -80,7 +84,45 @@ func NewPerson(store *Store, id string) (*Person, error) {
...
+
+	outbox, err := NewStream(store, id+"/outbox")
+	if err != nil {
+		return nil, errors.Wrap(err, "Creating outbox")
+	}
+
+	inbox.AttributedTo = jsonld.ToRef(id)
+	outbox.AttributedTo = jsonld.ToRef(id)
+	obj.Inbox = jsonld.ToRef(inbox.GetID())
+	obj.Outbox = jsonld.ToRef(outbox.GetID())
```
Comment (liclac) [inline models/entities/person.go]: Could we maybe make a `SetInbox()` and `SetOutbox` method pair for this? It feels like somewhere down the line, we will forget to set either `obj.{In,Out}box` or `{in,out}box.AttributedTo`.

File: models/entities/util.go [hunk 5]
```diff
@@ -0,0 +1,5 @@
+package entities
+
+func as2(postfix string) string {
```
Comment (liclac) [inline models/entities/util.go]: This really needs a better name :x

File: models/entities/person.go [hunk 6]
```diff
@@ -80,7 +84,45 @@ func NewPerson(store *Store, id string) (*Person, error) {
...
+
+	creation, err := NewActivity(store, outbox, []string{as2("Create")})
+	if err != nil {
+		return nil, errors.Wrap(err, "Creating Create activity")
+	}
+
+	creation.Actor = jsonld.ToRef(id)
+	creation.Obj = jsonld.ToRef(id)
+
+	if err := store.Save(inbox); err != nil {
```
Comment (liclac) [inline models/entities/person.go]: Thought: we could use `multierr` for this:

```
return obj, multierr.Combine(
  errors.Wrap(store.Save(inbox), "Saving inbox"),
  errors.Wrap(store.Save(outbox), "Saving outbox"),
  // ...
)
```

File: models/entities/stream.go [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+		ents[i] = o
+	}
+
+	m := map[string]interface{}{
+		"@list": ents,
+	}
+
+	o.(map[string]interface{})["https://www.w3.org/ns/activitystreams#items"] = []interface{}{m}
+
+	return jsonld.Compact(lib.GetHttpClient(ctx), o.(map[string]interface{}), "", "https://www.w3.org/ns/activitystreams")
```
Comment (liclac) [inline models/entities/stream.go]: Could we make a wrapper for this in `entities` that just takes a `context.Context` as its first parameter?

File: models/entities/stream.go [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+	}
+}
+
+func NewStream(store *Store, id string) (*Stream, error) {
+	obj := &Stream{
+		Object: Object{
+			Base: Base{
+				Meta: jsonld.Meta{},
+				ID:   id,
+				Type: []string{"https://www.w3.org/ns/activitystreams#Collection"},
```
Comment (liclac) [inline models/entities/stream.go]: What if we had a file that's just constants of every type in the AS spec, with namespace?

Comment (liclac) [reply models/entities/stream.go]: Somewhere down the line, we *are* going to put something in the wrong namespace by mistake, we already did once >_>

File: models/entities/activity.go (hunk 3 above)
Comment (erincandescent) [reply models/entities/activity.go]: In the code attached here (which creates the person's Create activity), I was putting the activity as a child of the outbox itself

so you end up with `example.com/~person/outbox/83428934789`

you would probably use the person as the parent of the note.
```
Object: example.com/~person/34789234789
Activity: example.com/~person/outbox/837489234789
```

File: models/entities/person.go (hunk 4 above)
Comment (erincandescent) [reply models/entities/person.go]: This sounds like a good idea

Actually I'm tempted to make accessors (GetInbox, GetOutbox?) which lazily create inboxes/outboxes?

File: models/entities/util.go (hunk 5 above)
Comment (erincandescent) [reply models/entities/util.go]: maybe put these functions in a ns package?

ns.AS2()
ns.LDP()

File: models/entities/stream.go (hunk 8 above)
Comment (erincandescent) [reply models/entities/stream.go]: I'm not sure how best to do that because AS2 types are case sensitive :/

File: models/entities/person.go (hunk 4 above)
Comment (liclac) [reply models/entities/person.go]: I like this idea, especially if you do both of these things

File: models/entities/stream.go (hunk 8 above)
Comment (liclac) [reply models/entities/stream.go]: Maybe something like `AS2_Object` vs `AS2_object`? It doesn't have to look very pretty, just prevent typos.

File: models/entities/util.go (hunk 5 above)
Comment (liclac) [reply models/entities/util.go]: Looks good! We could then use this in the typename constants as well.

File: models/entities/person.go (hunk 6 above)
Comment (liclac) [reply models/entities/person.go]: It feels silly to have this silly long list of `if err := ...; err != nil { return err }`! It's like half the function body >_<

**Pull request merged**